d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,177 @@
+      return;
+    }
+
+    cluster._uris = cluster._oldLb.Uris().size();
+    Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+    for (URI uri : cluster._oldLb.Uris()) {
+      if (!newLbUris.remove(uri)) {
+        continue;
+      }
+
+      if (compareURI(uri, cluster._oldLb, cluster._newLb)) {
+        cluster._matched++;
+      }
+    }
+    // add the remaining unmatched URIs in newLbUris to the uri count
+    cluster._uris += newLbUris.size();
+
+    if (cluster._matched != cluster._uris) {
+      RATE_LIMITED_LOGGER.info("Mismatched cluster properties for {} (match score: {}, total uris: {}):"
+              + "\nOld LB: {}\nNew LB: {}",
+          clusterName, cluster._matched / cluster._uris, cluster._uris, cluster._oldLb, cluster._newLb);
+    }
+
+    _totalUris += cluster._uris;
+    _matchedUris += cluster._matched;
+
+    updateJmxMetrics(clusterName, cluster);
+  }
+
+  private void updateJmxMetrics(String clusterName, ClusterMatchRecord cluster) {
+    _dualReadLoadBalancerJmx.setClusterMatchRecord(clusterName, cluster);
+    _dualReadLoadBalancerJmx.setUriPropertiesSimilarity((double) _matchedUris / (double) _totalUris);
+  }
+
+  private static boolean compareURI(URI uri, UriProperties oldLb, UriProperties newLb) {
+    String clusterName = oldLb.getClusterName();
+    return compareMaps("partition desc", clusterName, uri, UriProperties::getPartitionDesc, oldLb, newLb) &&
+        compareMaps("specific properties", clusterName, uri, UriProperties::getUriSpecificProperties, oldLb, newLb);
+  }
+
+  private static <K, V> boolean compareMaps(
+      String type, String cluster, URI uri, Function<UriProperties, Map<URI, Map<K, V>>> extractor,
+      UriProperties oldLb, UriProperties newLb
+  ) {
+    Map<K, V> oldData = extractor.apply(oldLb).get(uri);
+    Map<K, V> newData = extractor.apply(newLb).get(uri);
+    if (Objects.equals(oldData, newData)) {
+      return true;
+    }
+
+    LOG.debug("URI {} for {}/{} mismatched between old and new LB.\nOld LB: {}\nNew LB: {}",
+        type, cluster, uri, oldData, newData);
+    return false;
+  }
+
+  @VisibleForTesting
+  int getTotalUris() {
+    return _totalUris;
+  }
+
+  @VisibleForTesting
+  int getMatchedUris() {
+    return _matchedUris;
+  }
+
+  public static class ClusterMatchRecord {
+    @Nullable
+    @VisibleForTesting
+    UriProperties _oldLb;
+
+    @Nullable
+    @VisibleForTesting
+    UriProperties _newLb;
+
+    @VisibleForTesting
+    int _uris;
+
+    @VisibleForTesting
+    int _matched;
+
+    @Override
+    public String toString() {
+      return "ClusterMatchRecord{ " +
+          "\nTotal Uris: " + _uris + ", Matched: " + _matched +
+          "\nOld LB: " + _oldLb +
+          "\nNew LB: " + _newLb +
+          '}';
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (this == obj)
+        return true;
+      if (obj == null)
+        return false;
+      if (getClass() != obj.getClass())
+        return false;
```

Comment:
The go/javacode guidelines says:
> Always use braces around control statements

```suggestion
      if (this == obj) {
        return true;
      }
      if (obj == null) {
        return false;
      }
      if (getClass() != obj.getClass()) {
        return false;
      }
```

Reply:
updated.